TODO/beyond-browser-web-desktop-apps.md

```diff
@@ -442,31 +439,31 @@ describe('application launch', function () {
+这不仅仅是关乎桌面 app 或者 web app。作为一个 web 开发者，你可以用任何工具制作 app 确保在任何平台和环境中运行。但是为什么没有一劳永逸的办法呢？我们还需要努力，但这是值得的。接下来我会提到一些相关的话题和工具，考虑到它们太过复杂，我就点到为止。
 
-First of all, forget about “browser lock-in,” native WebSockets, etc. The same goes for ES6. You can either revert to writing plain old ES5 JavaScript or use something like [Babel](https://babeljs.io/) to transpile your ES6 into ES5, for web use.
+首先，忘记什么“浏览器限定”和原生 WebSockets 等等其他的事情。它们都提倡 ES6.你要么写纯粹的 ES5，要么用类似 [Babel](https://babeljs.io/) 的工具来把 ES6 代码编译成 ES5，供 web 使用。
 
-You also have `require`s throughout your code (for importing other scripts or modules), which a browser won’t understand. Use a module bundler that supports CommonJS (i.e. Node.js-style `require`s), such as [Rollup](http://rollupjs.org), [webpack](https://webpack.github.io) or [Browserify](http://browserify.org). When making a build for the web, a module bundler will run over your code, traverse all of the `require`s and bundle them up into one script for you.
+你的代码里也会写满了许多浏览器不会理解的`require`（用来引入其他脚本文件或者模块）。使用支持 CommonJS 的模块打包器，比如 [Rollup](http://rollupjs.org)，[webpack](https://webpack.github.io) 或者 [Browserify](http://browserify.org)。当构建 web app的时候，模块打包器会遍历代码，找到所有的`require`然后把他们放在一个脚本文件里。
 
-Any code using Node.js or Electron APIs (i.e. to write to disk or integrate with the desktop environment) should not be called when the app is running on the web. You can detect this by checking whether `process.version.nwjs` or `process.versions.electron` exists; if it does, then your app is currently running in the desktop environment.
+任何用到 Node.js 或者 Electron API（比如写盘操作或者整合桌面环境TODO）都不应该在 app 运行在 web 端的时候被调用。你可以通过检测`process.version.nwjs` 和 `process.versions.electron` 是否存在来侦测 API 的调用。如果存在，则表明 app 当前运行在桌面环境。
 
-Even then, you’ll be loading a lot of redundant code in the web app. Let’s say you have a `require` guarded behind a check like `if(app.isInDesktop)`, along with a big chunk of desktop-specific code. Instead of detecting the environment at runtime and setting `app.isInDesktop`, you could pass `true` or `false` into your app as a flag at buildtime (for example, using the [envify](https://github.com/hughsk/envify) transform for Browserify). This will aide your module bundler of choice when it’s doing its static analysis and tree-shaking (i.e. dead-code elimination). It will now know whether `app.isInDesktop` is `true`. So, if you’re running your web build, it won’t bother going inside that `if` statement or traversing the `require` in question.
+即便如此，你仍需要在 web app 上加载大量冗余代码。假设你的代码中`if(app.isInDesktop)` 后面紧接着和桌面环境有关的`require` 代码。并非在 app 运行的时候来检测当前运行环境，同时设置对应的 `app.isInDesktop`，你可以把 `true` 和 `false` 当做 flag 在构建的时候传值到 app。它会知道 `app.isInDesktop` 是否为 `true`。因此，当你运行 web app 的时候，它不会到代码里去找对应的 `if` 条件，或者找到相关的`require`。
```

Comment:
- 并非 -> 与其.....不如.....
- (例如，在 Browserify 中使用 [envify](https://github.com/hughsk/envify) )

> This will aide your module bundler of choice when it’s doing its static analysis and tree-shaking (i.e. dead-code elimination).

漏翻
> 在它进行静态和树状分析（也就是消除无用代码）时，这将有助于模块捆绑的选择。